### ISSUE TITLE
Wait API with and without notification pop-up

### ISSUE THREAD

**Issue by db0** [kellyelton/OCTGN#1365]
One of my games requires that both players simulataneously reveal their poker hand when they've finished redrawing it. I would like to add a function where a player presses ready and the system sets a shared variable and then waits until all other players have set their own variables as well. While it's waiting it would put a pop-up window with an option to cancel out of the loop (in case of errors). Finally It reveals all hands simultaneously once everyone's shared variable is set.

The wait API should handle both a wait delay with a pop-up window and one without (for more basic scripting usage)

It should be something like ```octgnWait(time = 1, notification = False)``` where the number is the seconds to wait, while the notification is a boolean which if set to True, pops-up a information window letting the player they're waiting for X seconds (counting down) and a button to abort the wait. The API should return True is the wait ended successfully, or False if it was aborted by the player.

**Comment by brine** [comment]
can't you do this with the OnPlayerGlobalVariableChanged events?

```python
def confirmHand(group, x = 0, y = 0):
    mute()
    setGlobalVariable("lockedHand", "True")
    notify("{} is happy with their hand.".format(me))

def OnPlayerGlobalVariableChanged(player, name, oldValue, newValue):
    mute()
    if name != "lockedHand":
        return ## we only care about the lockedHand variable
    if not me.isActivePlayer:
        return ## only the active player executes this event
    if newValue == "False":
        return ## skip if the hand wasn't confirmed
    for p in getPlayers():
        if getGlobalVariable("lockedHand"} == "False":
            return ## one or more players haven't confirmed their hands yet
    ## Do stuff if all players have confirmed their hands
    for p in getPlayers():
        remoteCall(p, "revealHands", [])

def revealHands():
    mute()
    setGlobalVariable("lockedHand", "False")
    doWhateverYouWant()
```

**Comment by db0** [comment]
I think i'd run into the same issue I'm running into currently. When both players reveal their hand simultaneously, the game ends up executing the "reveal hands" scripts twice.

**Comment by brine** [comment]
I have an isActivePlayer check in the event, so the trigger only processes once

**Comment by db0** [comment]
Smart, but my game doesn't use active players since phases are global for both players :-/

**Comment by kellyelton** [comment]
have it only run for the first player...if it's global it doesn't matter who runs it right?

**Comment by brine** [comment]
then make it so only the hosting player, `getPlayers()[0]` triggers it

**Comment by db0** [comment]
doesn't getPlayers() order differ per player?

**Comment by kellyelton** [comment]
player(1) is always the same player for everyone

**Comment by db0** [comment]
Sorry, you mean getPlayers()[0]? Or is this a different API?

**Comment by brine** [comment]
he meant getPlayers(), it returns the Player objects list sorted by Player._id, which are consistent between all clients.  So getPlayers()[0] is ALWAYS the hosting player, which is Player(1)

**Comment by db0** [comment]
Ah, didn't know that. I thought it worked like players[]. I'll try it out.

**Comment by kellyelton** [comment]
well it's less crazy to just to Player(1)

**Comment by brine** [comment]
yeah getPlayers() just returns [Player(1), Player(2), Player(3), ...], so `getPlayers()[0] == Player(1)`.  Kelly's right, if you just want to return the hosting player then Player(1) is cleaner

which is super confusing cuz there's the old `players[]` API call which returns that unsorted list of players, and is usually garbage

**Comment by kellyelton** [comment]
yeah....Player(1) 1 is the ID, not an index of an array...maybe that makes it clearer....the hosting players ID is 1.

**Comment by kellyelton** [comment]
Which you can get with 

    p = Player(1)
    notify("{}".format(p._id))

**Comment by brine** [comment]
ok so apparently I lied, getPlayers() returns the same list that players[] does

**Comment by kellyelton** [comment]
whew knew?

**Comment by db0** [comment]
Ah ok so getPlayers() is also unordered per player after all? and Player() with capital P is the ordered API right?

**Comment by kellyelton** [comment]
Player creates a plyer object with ID as the constructor

**Comment by brine** [comment]
the only difference between getPlayers() and players[] is that players[] is initialized on startup, and contains the original players in the game.  So if a player leaves the game then players[] will break cuz it contains an unknown player.

getPlayers() does the same list, but because it's a callable function it will return the players at that moment, so it'll ignore players who left the game

**Comment by kellyelton** [comment]
They're almost completely unrelated...getPlayers is a list of players based on you as the 0 player...so it's not unordered....Player is more like the Card object....Card(cardId)....Player(playerId)...or NamedObject(namedObjectId)

**Comment by db0** [comment]
Yeah I understood that. Just making sure because you initially wrote player with lowercase.

Do player IDs start at 0, or 1?

**Comment by brine** [comment]
1

I think 0 is reserved for the global player but that stuff's not really functional yet.

**Comment by kellyelton** [comment]
OH yeah that was a typo, player(1) isn't anything.

**Comment by kellyelton** [comment]
Also this kind of back and forth would be easier in a chat room instead of a github issue ;)

**Comment by db0** [comment]
No access to chat room from work :-/

**Comment by Gravecorp** [comment]
jeez stop blowing up my mailbox >_<
